yarpl/include/yarpl/observable/ObservableOperator.h

```diff
@@ -481,6 +481,73 @@ class SubscribeOnOperator
   std::unique_ptr<Worker> worker_;
 };
 
+template <typename T, typename GeneratorClass>
+class IteratorObservable : public Observable<T> {
+public:
+  Reference<Subscription> subscribe(Reference<Observer<T>> observer) override {
+    static_assert(
+      std::is_base_of<
+        IteratorObservable,
+        GeneratorClass
+      >::value,
+      "Inherit from IteratorObservable like `MyClass : public IteratorObservable<T, MyClass>`"
+    );
+
+    // can be stack allocated; we know nothing escapes
+    GeneratorClass instantiated(*static_cast<GeneratorClass const*>(this));
+    // so we can call protected methods on IteratorObservable even if they were
+    // made private in the subclass
+    IteratorObservable* withvtable = &instantiated;
```

Comment:
I wonder if the compiler is smart enough to devirtualize the method calls to `withvtable` - it should have enough information to know it always points to a concrete `GeneratorClass`. Will investigate with godbolt.